---
title: "Variants_alleles_merging_DH2417vDH2415"
author: "Jay Goddard"
date: "2024-04-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(readr)
library(tidyverse)
```

## Purpose

This is a script that looks to merge the individual sample variant calls (/snpeff/<sample.ann.tsv>, which include the variant alleles) with the "merged" full sample set variant call file against the reference (/merged_matrix/merged.ann.tsv)

## Context

This script will run over the DH2415 and DH2417 variant calls against reference genome DH2415, as generated by the "relocated" PA_SNPcalling pipeline in /dartfs-hpc/rc/lab/H/HoganD/GMBSR/PA_SNPcalling-master/ and stored in /dartfs-hpc/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/

This document is based on the pilot version of this workflow, made yesterday on DH2415/DH2417 calls against PA14. That project, including Rmd file and generated parsed variant files, is located at /dartfs-hpc/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variant_parsing_R_20240411 

## Strategy

The <sample>.ann.tsv files each contain a `CHROM` and `POS` column from the reference genome, followed by columns for the reference and "variant" alleles, and a long list of columns giving information and variant impact predictions (which we will not use here). 

We will need to use both of these columns, as this de novo genome is fragmented into multiple contigs, which were labeled as chromosomes by ther genomic variant analysis software used. Due to this, the original `join` functions from dplyr used on the DH2415/DH2417 callas against PA14, will need to be modified slightly to use two column values at once.

The merged.ann.tsv file also contains the same `CHROM` and `POS` column, much more information of the same vein as the individual sample files (which we will also not interact with here), and columns at the end for the read depth for each sample at each position.

Both fortunately and unfortunately for downstream analysis, this read depth at each position is shared whether the sample has the reference or alternate allele at each variant position. We can use this in future analyses to throw out "bad" calls ie. variant calls with read depth support at far below the average read depth for that strain's alignment across the reference genome, and particularly the average read depth of that strain's alignment at that close region around the variant call position to the reference genome.


I plan to use the merge function, or alternatively a similar set-up with a for loop, to "match" the variant call ref `CHROM` and `POS` in the merged.ann.tsv matrix, match that to the same `CHROM` and `POS` in each sample's variant call individual files, and then append the "alternate allele" value for that strain at that position in a new column of the merged.ann.tsv file, next to it's read depth at that position.

Given that the merged.ann.tsv matrix (generated by freebayes from all samples' individual alignment bam files to the reference) theoretically contains the `POS` for all variants from ALL samples against the reference (ie. not all <sample>.ann.tsv files will contain all `CHROM` and `POS` values in the merged.ann.tsv, but all `CHROM` and `POS` in the merged.ann.tsv matrix will be contained in at least one of the <sample>.ann.tsv files).

### Expected results

First, DH2415 is expected to have no variant calls against it's own reference genome. If we do see DH2415 variant calls in these results, manual investigation will be required to figure out the quality of these calls and possible reasons for their origin. This will happen OUTSIDE this script, as I want to keep this specific workflow standardized.

The expected end product will be a dataframe identical to the original merged.ann.tsv, with the information at the end for later analyses:
  1. columns for each sample's read depth at that position
  2. columns at each position, with each sample's "alternate allele" at each position
    a. any empty cell values in these alternate allele columns will be assumed to be positions where that strain contains the reference allele
    b. I may also append the `POS` in addition to the alternate allele in the first test of this, to make sure that the alleles being shared are actually
    c. I will also make a final "safety check" dataframe for each sample, that will contain any `POS`/alternate alleles that were in a sample's individual variant call file, but were not included in the merged.matrix.tsv. Hopefully this dataframe will be empty, but it will allow for a sanity-check that the merged.raw.vcf created by freebayes in the pipeline does not "miss" any variants that freebayes does detect/call in the individual <sample>.raw.vcf files.

```{r data-loading}
DH2415_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/snpeff/DH2415.ann.tsv', col_names = TRUE, skip = 1)
DH2417_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/snpeff/DH2417.ann.tsv', col_names = TRUE, skip = 1)
merged_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/merged_matrix/merged.ann.tsv', col_names = TRUE, skip = 0)
```

We are definitely seeing DH2415 variants called against itself. I am hopeful that these will all fall in intergenic regions, and in particular at the end of repetitive regions (or regions that share close homology in multiple parts of the genome, like tRNA sites).

So, one of my first manual checks later (possibly attached to the end of this script) will be counting the number and overall frequency of variant calls WITHIN CODING REGIONS in DH2415 against itself, and DH2417 against DH2417. I expect that DH2417 variants will have a much higher frequency, read depth, and average quality score than the DH2415 calls, and a higher number in defined coding regions. Tbd

## Looping through each sample dataframe to merge variant alleles into the merged matrix


```{r pressure, echo=FALSE}
merged2_df <- full_join(merged_df, DH2417_df, by = c("CHROM" = "CHROM", "POS" ="POS"), suffix = c("","_DH2417"))# %>%
  #full_join(DH2415_df, by = "POS", suffix = c("","_DH2415"))

merged3_df <- full_join(merged_df, DH2417_df, by = c("CHROM" = "CHROM", "POS" ="POS"), suffix = c("","_DH2417")) %>%
  full_join(DH2415_df, by = c("CHROM" = "CHROM", "POS" ="POS"), suffix = c("","_DH2415"))

#now, subsetting merged2_df and merged3_df:
merged4_df <- merged2_df[,c(1:58, 61:63, 70)]#this is only DH2417 calls against DH2415, just in case all DH2415 calls are low quality and blow up the full merged call dataframes
merged5_df <- merged3_df[,c(1:58, 61:63, 70, 124:126, 133)] #full dataset, including possible chaos introduced from DH2415 variant self-calls

merged6_df <- merged2_df[,c(1:58, 3:5, 62, 61,63, 70)] #DH2417 calls, parsed, reordered
merged7_df <- merged3_df[,c(1:58, 3:5, 62, 125, 61, 124, 63, 70,126, 133)] #full, parsed, reordered
#DH2415_solo_df <- #plan was to make reordered dataframe/file for DH2415_df, but that exists in the original tsv form
#DH2417_solo_df <- #plan was to make reordered dataframe/file for DH2417_df, but that exists in the original tsv form
```

## Checks for "lost" variant calls from individual samples, not included in the merged matrix variant call list

```{r}
DH2415_lost_df <- anti_join(DH2415_df, merged_df, c("CHROM" = "CHROM", "POS" ="POS"))
DH2417_lost_df <- anti_join(DH2417_df, merged_df,  c("CHROM" = "CHROM", "POS" ="POS"))
DH2415_lostfromm_df <- anti_join(merged_df, DH2415_df, c("CHROM" = "CHROM", "POS" ="POS"))
DH2417_lostfromm_df <- anti_join(merged_df, DH2417_df, c("CHROM" = "CHROM", "POS" ="POS"))
unique_inDH2415_vDH2417_df <- anti_join(DH2415_df, DH2417_df, c("CHROM" = "CHROM", "POS" ="POS"))
unique_inDH2417_vDH2415_df <- anti_join(DH2417_df, DH2415_df, c("CHROM" = "CHROM", "POS" ="POS"))
#I think I'm missing some comparisons here that I want to make, but I'll come back to this if I think of anything specific while analyzing what I have generated above
```

Now, saving these files for manual parsing/exploration:

```{r}
write_csv(merged2_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/DH2417-merged-variants.csv")
write_csv(merged3_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/DH2417&DH2415-merged-variants.csv")

write_csv(merged4_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/parsed-DH2417-merged-variants.csv")
write_csv(merged5_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/parsed-DH2417&DH2415-merged-variants.csv")

write_csv(merged6_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/parsed-ordered-DH2417-merged-variants.csv")
write_csv(merged7_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/parsed-ordered-DH2417&DH2415-merged-variants.csv")

write_csv(DH2415_lost_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/DH2415_lost.csv")
write_csv(DH2417_lost_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/DH2417_lost.csv")
write_csv(DH2415_lostfromm_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/DH2415_lost-from-merged.csv")
write_csv(DH2417_lostfromm_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/DH2417_lost-from-merged.csv")
write_csv(unique_inDH2415_vDH2417_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/unique_inDH2415_vDH2417.csv")
write_csv(unique_inDH2417_vDH2415_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2417vDH2415_20240411/variants_parsing_R_20240412/lost_variants/unique_inDH2417_vDH2415.csv")
```



