---
title: "Variants_alleles_merging_DH2415-DH2417"
author: "Jay Goddard"
date: "2024-04-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(readr)
library(tidyverse)
```

## Purpose

This is a script that looks to merge the individual sample variant calls (/snpeff/<sample.ann.tsv>, which include the variant alleles) with the "merged" full sample set variant call file against the reference (/merged_matrix/merged.ann.tsv)

## Context

This script will run over the DH2415 and DH2417 variant calls against reference genome PA14, as generated by the "relocated" PA_SNPcalling pipeline in /dartfs-hpc/rc/lab/H/HoganD/GMBSR/PA_SNPcalling-master/ and stored in /dartfs-hpc/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/

## Strategy

The <sample>.ann.tsv files each contain a `POS` column from the reference genome, followed by columns for the reference and "variant" alleles, and a long list of columns giving information and variant impact predictions (which we will not use here).

The merged.ann.tsv file also contains the same `POS` column, much more information of the same vein as the individual sample files (which we will also not interact with here), and columns at the end for the read depth for each sample at each position.

Both fortunately and unfortunately for downstream analysis, this read depth at each position is shared whether the sample has the reference or alternate allele at each variant position. We can use this in future analyses to throw out "bad" calls ie. variant calls with read depth support at far below the average read depth for that strain's alignment across the reference genome, and particularly the average read depth of that strain's alignment at that close region around the variant call position to the reference genome.


I plan to use the merge function, or alternatively a similar set-up with a for loop, to "match" the variant call ref `POS` in the merged.ann.tsv matrix, match that to the same `POS` in each sample's variant call individual files, and then append the "alternate allele" value for that strain at that position in a new column of the merged.ann.tsv file, next to it's read depth at that position.

Given that the merged.ann.tsv matrix (generated by freebayes from all samples' individual alignment bam files to the reference) theoretically contains the `POS` for all variants from ALL samples against the reference (ie. not all <sample>.ann.tsv files will contain all `POS` values in the merged.ann.tsv, but all `POS` in the merged.ann.tsv matrix will be contained in at least one of the <sample>.ann.tsv files).

### Expected results

The expected end product will be a dataframe identical to the original merged.ann.tsv, with the information at the end for later analyses:
  1. columns for each sample's read depth at that position
  2. columns at each position, with each sample's "alternate allele" at each position
    a. any empty cell values in these alternate allele columns will be assumed to be positions where that strain contains the reference allele
    b. I may also append the `POS` in addition to the alternate allele in the first test of this, to make sure that the alleles being shared are actually
    c. I will also make a final "safety check" dataframe for each sample, that will contain any `POS`/alternate alleles that were in a sample's individual variant call file, but were not included in the merged.matrix.tsv. Hopefully this dataframe will be empty, but it will allow for a sanity-check that the merged.raw.vcf created by freebayes in the pipeline does not "miss" any variants that freebayes does detect/call in the individual <sample>.raw.vcf files.

```{r data-loading}
DH2415_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/snpeff/DH2415.ann.tsv', col_names = TRUE, skip = 1)
DH2417_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/snpeff/DH2417.ann.tsv', col_names = TRUE, skip = 1)
merged_df <- read_tsv('/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/merged_matrix/merged.ann.tsv', col_names = TRUE, skip = 0)
```

## Looping through each sample dataframe to merge variant alleles into the merged matrix


```{r pressure, echo=FALSE}
merged_appended_df <- merge(merged_df, DH2415_df, by='POS') #this did not do what we wanted. 1. it combined ALL columns from sample dataframe, and 2. it excluded any positions in merged matrix that this sample in particular did not have (basically, the central component of a venn diagram wasthis output)

merged2_df <- merged_df %>% full_join(DH2415_df) #! Can't join `x$ID` with `y$ID` due to incompatible types., ℹ `x$ID` is a <double>., ℹ `y$ID` is a <character>.
merged3_df <- merged_df %>% bind_rows(DH2415_df)#! Can't combine `..1$ID` <double> and `..2$ID` <character>.
merged4_df <- merged_df %>% full_join(DH2415_df, by = c("POS" ="POS"))# definitely closer to what we want
merged5_df <- merged_df %>% inner_join(DH2415_df, 
                                      by = c("POS" ="POS"),
                                      suffix = c("_m","_DH2415"))#this one excluded any rows not SHARED
merged6_df <- full_join(merged_df, DH2415_df, by = c("POS" ="POS"), suffix = c("","_DH2415"))
merged7_df <- full_join(merged_df, DH2417_df, by = c("POS" ="POS"), suffix = c("","_DH2417")) %>%
  full_join(DH2415_df, by = "POS", suffix = c("_m","_DH2415"))
merged8_df <- full_join(merged_df, DH2417_df, by = c("POS" ="POS"), suffix = c("","_DH2417")) %>%
  full_join(DH2415_df, by = "POS", suffix = c("","_DH2415"))

#now, subsetting merged8_df:
merged9_df <- merged8_df[,c(1:58, 61:63, 70, 124:126, 133)]
merged10_df <- merged8_df[,c(1:58, 4:5, 62, 125, 61, 124, 63, 70,126, 133)]
```

## Checks for "lost" variant calls from individual samples, not included in the merged matrix variant call list

```{r}
DH2415_lost_df <- anti_join(DH2415_df, merged_df, by ="POS")
DH2417_lost_df <- anti_join(DH2417_df, merged_df,  by ="POS")
DH2415_lostfromm_df <- anti_join(merged_df, DH2415_df, by ="POS")
DH2417_lostfromm_df <- anti_join(merged_df, DH2417_df, by ="POS")
unique_inDH2415_vDH2417_df <- anti_join(DH2415_df, DH2417_df, by ="POS")
unique_inDH2417_vDH2415_df <- anti_join(DH2417_df, DH2415_df, by ="POS")
```

Now, saving these files for manual parsing/exploration:

```{r}
write_csv(merged8_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/full-merged-variants.csv")
write_csv(merged9_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/parsed-merged-variants.csv")
write_csv(merged10_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/parsed-ordered-merged-variants.csv")

write_csv(DH2415_lost_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/DH2415_lost.csv")
write_csv(DH2417_lost_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/DH2417_lost.csv")
write_csv(DH2415_lostfromm_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/DH2415_lost-from-merged.csv")
write_csv(DH2417_lostfromm_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/DH2417_lost-from-merged.csv")
write_csv(unique_inDH2415_vDH2417_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/unique_inDH2415_vDH2417.csv")
write_csv(unique_inDH2417_vDH2415_df,"/Volumes/rc/lab/H/HoganD/Genomic_analyses/DH2415_DH2417_newsetup_20240408/variants_parsing_R_20240411/lost_variants/unique_inDH2417_vDH2415.csv")
```



